File: services/data-search/internal/infrastructure/events/retry_topology.go
package events

import (
	"context"
	"math/rand"
	"strconv"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

const (
	MaxRetries  = 3
	baseDelayMs = 1000
)

// RetryQueueName returns the retry queue name for a given source queue.
func RetryQueueName(sourceQueue string) string {
	return sourceQueue + ".retry"
}

// AssertRetryQueue declares the retry queue with a dead-letter back to the
// main exchange so expired messages are redelivered to the source queue.
func AssertRetryQueue(ch *amqp.Channel, sourceQueue, sourceExchange, sourceRoutingKey string) error {
	_, err := ch.QueueDeclare(
		RetryQueueName(sourceQueue),
		true,  // durable
		false, // auto-delete
		false, // exclusive
		false, // no-wait
		amqp.Table{
			"x-dead-letter-exchange":    sourceExchange,
			"x-dead-letter-routing-key": sourceRoutingKey,
		},
	)
	return err
}

func getRetryCount(msg amqp.Delivery) int {
	if v, ok := msg.Headers["x-retry-count"].(int32); ok {
		return int(v)
	}
	if v, ok := msg.Headers["x-retry-count"].(int64); ok {
		return int(v)
	}
	return 0
}

func computeBackoffMs(count int) int {
	base := baseDelayMs * (1 << count) // 2^count * 1000
	jitter := 0.9 + rand.New(rand.NewSource(time.Now().UnixNano())).Float64()*0.2
	return int(float64(base) * jitter)
}

// ScheduleRetry publishes the message to the retry queue with an exponential
// backoff TTL. Returns false if max retries are exhausted — caller must nack
// to route the message to DLQ.
func ScheduleRetry(ch *amqp.Channel, msg amqp.Delivery, sourceQueue string) (bool, error) {
	count := getRetryCount(msg)
	if count >= MaxRetries {
		return false, nil
	}

	delay := computeBackoffMs(count)

	headers := amqp.Table{}
	for k, v := range msg.Headers {
		headers[k] = v
	}
	headers["x-retry-count"] = int32(count + 1)

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	err := ch.PublishWithContext(ctx, "", RetryQueueName(sourceQueue), false, false, amqp.Publishing{
		ContentType:  msg.ContentType,
		Body:         msg.Body,
		DeliveryMode: amqp.Persistent,
		Headers:      headers,
		Expiration:   strconv.Itoa(delay),
	})
	if err != nil {
		return false, err
	}

	return true, nil
}

File: services/process-data-ia/src/events/retry-topology.ts
import type { Channel, ConsumeMessage } from "amqplib";

export const MAX_RETRIES = 3;
const BASE_DELAY_MS = 1000;

export function getRetryCount(msg: ConsumeMessage): number {
  const raw = msg.properties.headers?.["x-retry-count"];
  return typeof raw === "number" ? raw : 0;
}

export function computeBackoffMs(retryCount: number): number {
  const base = BASE_DELAY_MS * 2 ** retryCount;
  const jitter = 0.9 + Math.random() * 0.2;
  return Math.round(base * jitter);
}

export async function assertRetryQueueFor(
  channel: Channel,
  sourceQueue: string,
  sourceExchange: string,
  sourceRoutingKey: string
): Promise<void> {
  const retryQueue = `${sourceQueue}.retry`;
  await channel.assertQueue(retryQueue, {
    durable: true,
    arguments: {
      "x-dead-letter-exchange": sourceExchange,
      "x-dead-letter-routing-key": sourceRoutingKey,
    },
  });
}

/**
 * Publica a mensagem na retry queue com TTL de backoff exponencial.
 * Retorna true se o retry foi agendado; false se max retries esgotadas
 * (o caller deve rotear para DLQ via nack).
 */
export function scheduleRetry(
  channel: Channel,
  msg: ConsumeMessage,
  sourceQueue: string
): boolean {
  const currentCount = getRetryCount(msg);
  if (currentCount >= MAX_RETRIES) {
    return false;
  }
  const delay = computeBackoffMs(currentCount);
  const retryQueue = `${sourceQueue}.retry`;
  channel.sendToQueue(retryQueue, msg.content, {
    persistent: true,
    expiration: String(delay),
    headers: {
      ...msg.properties.headers,
      "x-retry-count": currentCount + 1,
    },
  });
  return true;
}
